{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a72d1a4b",
   "metadata": {},
   "source": [
    "# Basic Search"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "143f2e20",
   "metadata": {},
   "source": [
    "The CDA provides a custom python tool for searching CDA data. [`Q`](usage/#q) (short for Query) offers several ways to search and filter data, and several input modes:\n",
    "\n",
    "---\n",
    "- **<a href=\"../../QuickStart/usage/#q\">Q.()</a>** builds a query that can be used by `run()` or `count()`\n",
    "- **<a href=\"../../QuickStart/usage/#qrun\">Q.run()</a>** returns data for the specified search \n",
    "- **<a href=\"../../QuickStart/usage/#qcount\">Q.count()</a>** returns summary information (counts) data that fit the specified search\n",
    "- **<a href=\"../../QuickStart/usage/#columns\">columns()</a>** returns entity field names\n",
    "- **<a href=\"../../QuickStart/usage/#unique_terms\">unique_terms()</a>** returns entity field contents\n",
    "\n",
    "---\n",
    "                                                                    \n",
    "Before we do any work, we needs to import these functions cdapython.\n",
    "We're also telling cdapython to report it's version so we can be sure we're using the one we mean to:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5245b7b0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2022.6.28\n"
     ]
    }
   ],
   "source": [
    "from cdapython import Q, columns, unique_terms, query\n",
    "print(Q.get_version())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d538da92",
   "metadata": {},
   "source": [
    "<div class=\"cdanote\" style=\"background-color:#b3e5d5;color:black;padding:20px;\">\n",
    "    \n",
    "CDA data comes from three sources:\n",
    "<ul>\n",
    "<li><b>The <a href=\"https://proteomic.datacommons.cancer.gov/pdc/\"> Proteomic Data Commons</a> (PDC)</b></li>\n",
    "<li><b>The <a href=\"https://gdc.cancer.gov/\">Genomic Data Commons</a> (GDC)</b></li>\n",
    "<li><b>The <a href=\"https://datacommons.cancer.gov/repository/imaging-data-commons\">Imaging Data Commons</a> (IDC)</b></li>\n",
    "</ul> \n",
    "    \n",
    "The CDA makes this data searchable in four main endpoints:\n",
    "\n",
    "<ul>\n",
    "<li><b>subject:</b> A patient entity captures the study-independent metadata for research subjects. Human research subjects are usually not traceable to a particular person to protect the subjects privacy.</li>\n",
    "<li><b>researchsubject:</b> A research subject is the entity of interest in a specific research study or project, typically a human being or an animal, but can also be a device, group of humans or animals, or a tissue sample. Human research subjects are usually not traceable to a particular person to protect the subjects privacy. This entity plays the role of the case_id in existing data. An individual who participates in 3 studies will have 3 researchsubject IDs</li>\n",
    "<li><b>specimen:</b> Any material taken as a sample from a biological entity (living or dead), or from a physical object or the environment. Specimens are usually collected as an example of their kind, often for use in some investigation.</li>\n",
    "<li><b>file:</b> A unit of data about subjects, researchsubjects, specimens, or their associated information</li>\n",
    "</ul>\n",
    "and two endpoints that offer deeper information about data in the researchsubject endpoint:\n",
    "<ul>\n",
    "<li><b>diagnosis:</b> A collection of characteristics that describe an abnormal condition of the body as assessed at a point in time. May be used to capture information about neoplastic and non-neoplastic conditions.</li>\n",
    "<li><b>treatment:</b> Represent medication administration or other treatment types.</li>\n",
    "</ul>\n",
    "Any metadata field can be searched from any endpoint, the only difference between search types is what type of data is returned by default. This means that you can think of the CDA as a really, really enormous spreadsheet full of data. To search this enormous spreadsheet, you'd want select columns, and then filter rows.\n",
    "</div>\n",
    "\n",
    "\n",
    "If you are looking to build a cohort of distinct individuals who meet some criteria, search by `subject`. If you want to build a cohort, but are particularly interested in studies rather than the participates per se, search by `researchsubject`. If you are looking for biosamples that can be ordered or a specfic format of information (for e.g. histological slides) start with `specimen`. If you are primarily looking for files you can reuse for your own analysis, start with `file`.\n",
    "\n",
    "In CDA search, these concepts can also be chained together, so you can look specifically for specimen subjects, or researchsubject diagnoses. In the four 'main' tables, all of the rows will have one or more files associated with them that can be directly found by chaining, as in specimen files. Diagnosis and treatment do not have files directly associated with them and so can only be used to find files in conjunction with the other searches.\n",
    "\n",
    "In all cases, any search can use any metadata field, the only difference between search types is what type of data you return by default. \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f76026a3",
   "metadata": {},
   "source": [
    "## Basic search with endpoints\n",
    "\n",
    "Let's try a broad search of the CDA to see what information exists about cancers that were first diagnosed in the brain. To run this simple search, we would first construct a query in `Q` and save it to a variable `myquery`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c92a98ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "myquery = Q('ResearchSubject.primary_diagnosis_site = \"brain\"')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e855573",
   "metadata": {
    "tags": [
     "Help!"
    ]
   },
   "source": [
    "\n",
    "<div class=\"cdawarn\" style=\"background-color:#f9cfbf;color:black;padding:20px;\">\n",
    "    \n",
    "<h3>Where did those terms come from?</h3>\n",
    "    \n",
    "If you aren't sure how we knew what terms to put in our search, please refer back to the <a href=\"../SearchTerms\">What search terms are available?</a> notebook. \n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d483e944",
   "metadata": {},
   "source": [
    "### subject\n",
    "Now we can use that query to search any of information types. Let's start by looking at what Subjects meet our criteria. To do that, we will send our query to the subject endpoint, then ask for it to run:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6d9137aa",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3566 ms\n"
     ]
    }
   ],
   "source": [
    "subjectresults = myquery.subject.run()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fdf7fb60",
   "metadata": {},
   "source": [
    "We saved the output in a variable `subjectresults`, so we don't get much visible output. To see what our results are, we need to look into the variable. The simplest way is to call `subjectresults` directly:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "cff572d7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\n",
       "            QueryID: 77faf08e-aeb1-4c83-aa00-0288493026e8\n",
       "            \n",
       "            Offset: 0\n",
       "            Count: 100\n",
       "            Total Row Count: 2384\n",
       "            More pages: True\n",
       "            "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subjectresults"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5186036c",
   "metadata": {},
   "source": [
    "This output tells us our QueryID, which we don't really need, but the computer does to track our questions. Then it tells us four parameters that describe our results:\n",
    "\n",
    "---\n",
    "\n",
    "- **Offset:** This is how many rows of information we've told the query to skip in the data, here we didn't tell it to skip anything, so the offset is zero\n",
    "- **Count:** This is how many rows the current page of our results table has. To keep searches fast, we default to pages with 100 rows.\n",
    "- **Total Row Count:** This is how many rows are in the full results table\n",
    "- **More pages:** This is always a True or False. False means that our current page has all the available results. True means that we will see only the first 100 results in this table, and will need to page through for more.\n",
    "\n",
    "---\n",
    "    \n",
    "We have 2384 subject matches, let's look at the actual table. The easiest way to do this is by using the python function `.to_dataframe()` on our `subjectresults` variable:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e97e19f4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>species</th>\n",
       "      <th>sex</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>days_to_birth</th>\n",
       "      <th>subject_associated_project</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>days_to_death</th>\n",
       "      <th>cause_of_death</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C3L-03588</td>\n",
       "      <td>[{'system': 'IDC', 'value': 'C3L-03588'}]</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[cptac_gbm]</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C3N-03810</td>\n",
       "      <td>[{'system': 'IDC', 'value': 'C3N-03810'}]</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[cptac_gbm]</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C466293</td>\n",
       "      <td>[{'system': 'PDC', 'value': 'C466293'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Proteogenomic Analysis of Pediatric Brain Can...</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Reported</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GENIE-JHU-03295</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'GENIE-JHU-03295'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>black or african american</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[GENIE-JHU]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GENIE-MDA-3866</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'GENIE-MDA-3866'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>white</td>\n",
       "      <td>not reported</td>\n",
       "      <td>-12053.0</td>\n",
       "      <td>[GENIE-MDA]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>TCGA-06-0149</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-06-0149'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-27315.0</td>\n",
       "      <td>[tcga_gbm, TCGA-GBM]</td>\n",
       "      <td>Dead</td>\n",
       "      <td>262.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>TCGA-06-2557</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-06-2557'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>black or african american</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-27980.0</td>\n",
       "      <td>[tcga_gbm, TCGA-GBM]</td>\n",
       "      <td>Dead</td>\n",
       "      <td>33.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>TCGA-08-0347</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-08-0347'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>not reported</td>\n",
       "      <td>not reported</td>\n",
       "      <td>-18384.0</td>\n",
       "      <td>[tcga_gbm, TCGA-GBM]</td>\n",
       "      <td>Dead</td>\n",
       "      <td>782.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>TCGA-08-0354</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-08-0354'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-19199.0</td>\n",
       "      <td>[tcga_gbm, TCGA-GBM]</td>\n",
       "      <td>Dead</td>\n",
       "      <td>546.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>TCGA-12-1089</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-12-1089'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-23419.0</td>\n",
       "      <td>[tcga_gbm, TCGA-GBM]</td>\n",
       "      <td>Dead</td>\n",
       "      <td>177.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 id                                         identifier  \\\n",
       "0         C3L-03588          [{'system': 'IDC', 'value': 'C3L-03588'}]   \n",
       "1         C3N-03810          [{'system': 'IDC', 'value': 'C3N-03810'}]   \n",
       "2           C466293            [{'system': 'PDC', 'value': 'C466293'}]   \n",
       "3   GENIE-JHU-03295    [{'system': 'GDC', 'value': 'GENIE-JHU-03295'}]   \n",
       "4    GENIE-MDA-3866     [{'system': 'GDC', 'value': 'GENIE-MDA-3866'}]   \n",
       "..              ...                                                ...   \n",
       "95     TCGA-06-0149  [{'system': 'GDC', 'value': 'TCGA-06-0149'}, {...   \n",
       "96     TCGA-06-2557  [{'system': 'GDC', 'value': 'TCGA-06-2557'}, {...   \n",
       "97     TCGA-08-0347  [{'system': 'GDC', 'value': 'TCGA-08-0347'}, {...   \n",
       "98     TCGA-08-0354  [{'system': 'GDC', 'value': 'TCGA-08-0354'}, {...   \n",
       "99     TCGA-12-1089  [{'system': 'GDC', 'value': 'TCGA-12-1089'}, {...   \n",
       "\n",
       "         species     sex                       race               ethnicity  \\\n",
       "0   Homo sapiens    None                       None                    None   \n",
       "1   Homo sapiens    None                       None                    None   \n",
       "2   homo sapiens    male                      white  not hispanic or latino   \n",
       "3   homo sapiens  female  black or african american                 Unknown   \n",
       "4   homo sapiens    male                      white            not reported   \n",
       "..           ...     ...                        ...                     ...   \n",
       "95  homo sapiens  female                      white  not hispanic or latino   \n",
       "96  homo sapiens    male  black or african american  not hispanic or latino   \n",
       "97  homo sapiens    male               not reported            not reported   \n",
       "98  homo sapiens  female                      white  not hispanic or latino   \n",
       "99  homo sapiens    male                      white  not hispanic or latino   \n",
       "\n",
       "    days_to_birth                         subject_associated_project  \\\n",
       "0             NaN                                        [cptac_gbm]   \n",
       "1             NaN                                        [cptac_gbm]   \n",
       "2             NaN  [Proteogenomic Analysis of Pediatric Brain Can...   \n",
       "3             NaN                                        [GENIE-JHU]   \n",
       "4        -12053.0                                        [GENIE-MDA]   \n",
       "..            ...                                                ...   \n",
       "95       -27315.0                               [tcga_gbm, TCGA-GBM]   \n",
       "96       -27980.0                               [tcga_gbm, TCGA-GBM]   \n",
       "97       -18384.0                               [tcga_gbm, TCGA-GBM]   \n",
       "98       -19199.0                               [tcga_gbm, TCGA-GBM]   \n",
       "99       -23419.0                               [tcga_gbm, TCGA-GBM]   \n",
       "\n",
       "    vital_status  days_to_death cause_of_death  \n",
       "0           None            NaN           None  \n",
       "1           None            NaN           None  \n",
       "2          Alive            NaN   Not Reported  \n",
       "3   Not Reported            NaN           None  \n",
       "4   Not Reported            NaN           None  \n",
       "..           ...            ...            ...  \n",
       "95          Dead          262.0           None  \n",
       "96          Dead           33.0           None  \n",
       "97          Dead          782.0           None  \n",
       "98          Dead          546.0           None  \n",
       "99          Dead          177.0           None  \n",
       "\n",
       "[100 rows x 11 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subjectresults.to_dataframe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e44a65df",
   "metadata": {},
   "source": [
    "By default `to_dataframe()` shows us the first and last five rows for the first page of our results, so we can easily preview our data.\n",
    "\n",
    "Since we queried the Subject endpoint, our default results tell us Subject level information, that is, information about unique individuals: their sex, race, age, species, etc. The `id` column tells us the unique identifier for each individual. The identifier column has nested information about what study or studies a Subject participated in, and will list all of their researchsubject identifiers. \n",
    "\n",
    "The `to_dataframe()` function converts the results to a pandas dataframe. So if we save the dataframe to a variable, we can use any pandas functions to explore it. For example, lets see whether any of the Subjects in our first 100 results are black or african american. First we'll save the results to a dataframe, then subset that dataframe to only show rows where the word \"black\" appears in the \"race\" column. \"NAs\" which are shown as \"None\" in these tables, so for our filter to work, we'll need to specifically tell it to ignore NAs:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4975c1e5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>species</th>\n",
       "      <th>sex</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>days_to_birth</th>\n",
       "      <th>subject_associated_project</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>days_to_death</th>\n",
       "      <th>cause_of_death</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GENIE-JHU-03295</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'GENIE-JHU-03295'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>black or african american</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[GENIE-JHU]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>TCGA-DU-A5TW</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-DU-A5TW'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>black or african american</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-12107.0</td>\n",
       "      <td>[TCGA-LGG, tcga_lgg]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>TCGA-14-3477</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-14-3477'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>black or african american</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-14082.0</td>\n",
       "      <td>[tcga_gbm, TCGA-GBM]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>TCGA-06-2557</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-06-2557'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>black or african american</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-27980.0</td>\n",
       "      <td>[tcga_gbm, TCGA-GBM]</td>\n",
       "      <td>Dead</td>\n",
       "      <td>33.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 id                                         identifier  \\\n",
       "3   GENIE-JHU-03295    [{'system': 'GDC', 'value': 'GENIE-JHU-03295'}]   \n",
       "40     TCGA-DU-A5TW  [{'system': 'GDC', 'value': 'TCGA-DU-A5TW'}, {...   \n",
       "72     TCGA-14-3477  [{'system': 'GDC', 'value': 'TCGA-14-3477'}, {...   \n",
       "96     TCGA-06-2557  [{'system': 'GDC', 'value': 'TCGA-06-2557'}, {...   \n",
       "\n",
       "         species     sex                       race               ethnicity  \\\n",
       "3   homo sapiens  female  black or african american                 Unknown   \n",
       "40  homo sapiens  female  black or african american  not hispanic or latino   \n",
       "72  homo sapiens  female  black or african american  not hispanic or latino   \n",
       "96  homo sapiens    male  black or african american  not hispanic or latino   \n",
       "\n",
       "    days_to_birth subject_associated_project  vital_status  days_to_death  \\\n",
       "3             NaN                [GENIE-JHU]  Not Reported            NaN   \n",
       "40       -12107.0       [TCGA-LGG, tcga_lgg]         Alive            NaN   \n",
       "72       -14082.0       [tcga_gbm, TCGA-GBM]         Alive            NaN   \n",
       "96       -27980.0       [tcga_gbm, TCGA-GBM]          Dead           33.0   \n",
       "\n",
       "   cause_of_death  \n",
       "3            None  \n",
       "40           None  \n",
       "72           None  \n",
       "96           None  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subjectdata = subjectresults.to_dataframe()\n",
    "subjectdata[subjectdata['race'].str.contains(\"black\", na=False)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8750617",
   "metadata": {},
   "source": [
    "There are subjects in our first hundred results that meet the criteria. If we just want to be sure that the data contains some value, this might be good enough. But often we want to search the entire set of results and not just the first page. \n",
    "\n",
    "We'll cover how to work with large results dataframes in the <a href=\"../Pagination\">Pagination</a> notebook. Or, learn how to get summary information from search results in the <a href=\"../DataSummaries\">Data Summaries</a> notebook."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "05e52f3f",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "<div class=\"cdadefine\" style=\"background-color:#add9e5;color:black;padding:20px;\">\n",
    "\n",
    "<h3>Subject Field Definitions</h3>\n",
    "\n",
    "<i>A patient entity captures the study-independent metadata for research subjects. Human research subjects are usually not traceable to a particular person to protect the subjects privacy.</i>\n",
    "\n",
    "    \n",
    "<ul>\n",
    "<li><b>id:</b> The 'logical' identifier of the entity in the system of record, e.g. a UUID. This 'id' is unique within a given system. The identified entity may have a different 'id' in a different system.\",STRING</li>\n",
    "<li><b>identifier:</b> A 'business' identifier for the entity, typically as provided by an external system or authority, that persists across implementing systems (i.e. a 'logical' identifier). Uses a specialized, complex 'Identifier' data type to capture information about the source of the business identifier - or a URI expressed as a string to an existing entity.</li>\n",
    "<li><b>identifier.system:</b> The system or namespace that defines the identifier.</li>\n",
    "<li><b>identifier.value:</b> The value of the identifier, as defined by the system.</li>\n",
    "<li><b>species:</b> The taxonomic group (e.g. species) of the subject. For MVP, since taxonomy vocabulary is consistent between GDC and PDC, using text. Ultimately, this will be a term returned by the vocabulary service.</li>\n",
    "<li><b>sex:</b> The biologic character or quality that distinguishes male and female from one another as expressed by analysis of the person's gonadal, morphologic (internal and external), chromosomal, and hormonal characteristics.</li>\n",
    "<li><b>race:</b> An arbitrary classification of a taxonomic group that is a division of a species. It usually arises as a consequence of geographical isolation within a species and is characterized by shared heredity, physical attributes and behavior, and in the case of humans, by common history, nationality, or geographic distribution. The provided values are based on the categories defined by the U.S. Office of Management and Business and used by the U.S. Census Bureau.</li>\n",
    "<li><b>ethnicity:</b> An individual's self-described social and cultural grouping, specifically whether an individual describes themselves as Hispanic or Latino. The provided values are based on the categories defined by the U.S. Office of Management and Business and used by the U.S. Census Bureau.</li>\n",
    "<li><b>days_to_birth:</b> Number of days between the date used for index and the date from a person's date of birth represented as a calculated negative number of days.</li>\n",
    "<li><b>subject_associated_project:</b> The list of Projects associated with the Subject.</li>\n",
    "<li><b>vital_status:</b> Coded value indicating the state or condition of being living or deceased; also includes the case where the vital status is unknown.</li>\n",
    "<li><b>days_to_death:</b> Number of days between the date used for index and the date from a person's date of death represented as a calculated number of days.</li>\n",
    "<li><b>cause_of_death:</b> Coded value indicating the circumstance or condition that results in the death of the subject.</li>\n",
    "</ul>  \n",
    "\n",
    "</div>\n",
    "    \n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73745aa6",
   "metadata": {},
   "source": [
    "### researchsubject\n",
    "\n",
    "If we're interested in what researchsubjects meet our critera, we can also run our query against the researchsubject endpoint:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a2d8c874",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3696 ms\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "            QueryID: 954e86eb-240e-40eb-a3f3-d33cb4a22c83\n",
       "            \n",
       "            Offset: 0\n",
       "            Count: 100\n",
       "            Total Row Count: 3716\n",
       "            More pages: True\n",
       "            "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "researchsubjectresults = myquery.researchsubject.run()\n",
    "researchsubjectresults"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "977f36fe",
   "metadata": {},
   "source": [
    "Now we see that our 2384 subjects have 3716 researchsubjects between them, that means that some, but not all, of our subjects were participants in more than one study. Let's peek at the data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a20f7e62",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>member_of_research_project</th>\n",
       "      <th>primary_diagnosis_condition</th>\n",
       "      <th>primary_diagnosis_site</th>\n",
       "      <th>subject_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>54ada9c7-e8f7-4ecc-b829-903be7e3ad82</td>\n",
       "      <td>[{'system': 'GDC', 'value': '54ada9c7-e8f7-4ec...</td>\n",
       "      <td>GENIE-NKI</td>\n",
       "      <td>Neoplasms, NOS</td>\n",
       "      <td>Brain</td>\n",
       "      <td>GENIE-NKI-8IN3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8232489b-f2dc-44a1-b6dc-38b94bda584b</td>\n",
       "      <td>[{'system': 'GDC', 'value': '8232489b-f2dc-44a...</td>\n",
       "      <td>GENIE-JHU</td>\n",
       "      <td>Neoplasms, NOS</td>\n",
       "      <td>Brain</td>\n",
       "      <td>GENIE-JHU-01660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>b936c96e-70ab-46eb-a87f-73290c43f288</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'b936c96e-70ab-46e...</td>\n",
       "      <td>GENIE-NKI</td>\n",
       "      <td>Neoplasms, NOS</td>\n",
       "      <td>Brain</td>\n",
       "      <td>GENIE-NKI-U4QV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>be68cfb0-c47f-49a1-9e39-b0b3e57372fc</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'be68cfb0-c47f-49a...</td>\n",
       "      <td>GENIE-JHU</td>\n",
       "      <td>Neoplasms, NOS</td>\n",
       "      <td>Brain</td>\n",
       "      <td>GENIE-JHU-02615</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>99b34e59-a5d0-4a8b-b777-3a592c5bea1b</td>\n",
       "      <td>[{'system': 'GDC', 'value': '99b34e59-a5d0-4a8...</td>\n",
       "      <td>GENIE-JHU</td>\n",
       "      <td>Neoplasms, NOS</td>\n",
       "      <td>Brain</td>\n",
       "      <td>GENIE-JHU-02402</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>6acf365b-19ce-4c97-92dd-865540a47e67</td>\n",
       "      <td>[{'system': 'GDC', 'value': '6acf365b-19ce-4c9...</td>\n",
       "      <td>CPTAC-3</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>Brain</td>\n",
       "      <td>C3L-02642</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>94f90e9a-ecfe-4682-8d19-2ac88b31a5b9</td>\n",
       "      <td>[{'system': 'GDC', 'value': '94f90e9a-ecfe-468...</td>\n",
       "      <td>CPTAC-3</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>Brain</td>\n",
       "      <td>C3N-01334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>e1bd6dfa-0567-46e7-991c-63eefaf5bd7a</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'e1bd6dfa-0567-46e...</td>\n",
       "      <td>CPTAC-3</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>Brain</td>\n",
       "      <td>C3L-00674</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>142a1357-f1e8-40b9-82cf-f29577058598</td>\n",
       "      <td>[{'system': 'GDC', 'value': '142a1357-f1e8-40b...</td>\n",
       "      <td>CPTAC-3</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>Brain</td>\n",
       "      <td>C3N-01852</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>d86c3d4a-861b-4211-9bcd-c9669fcc480b</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'd86c3d4a-861b-421...</td>\n",
       "      <td>CPTAC-3</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>Brain</td>\n",
       "      <td>C3L-03260</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      id  \\\n",
       "0   54ada9c7-e8f7-4ecc-b829-903be7e3ad82   \n",
       "1   8232489b-f2dc-44a1-b6dc-38b94bda584b   \n",
       "2   b936c96e-70ab-46eb-a87f-73290c43f288   \n",
       "3   be68cfb0-c47f-49a1-9e39-b0b3e57372fc   \n",
       "4   99b34e59-a5d0-4a8b-b777-3a592c5bea1b   \n",
       "..                                   ...   \n",
       "95  6acf365b-19ce-4c97-92dd-865540a47e67   \n",
       "96  94f90e9a-ecfe-4682-8d19-2ac88b31a5b9   \n",
       "97  e1bd6dfa-0567-46e7-991c-63eefaf5bd7a   \n",
       "98  142a1357-f1e8-40b9-82cf-f29577058598   \n",
       "99  d86c3d4a-861b-4211-9bcd-c9669fcc480b   \n",
       "\n",
       "                                           identifier  \\\n",
       "0   [{'system': 'GDC', 'value': '54ada9c7-e8f7-4ec...   \n",
       "1   [{'system': 'GDC', 'value': '8232489b-f2dc-44a...   \n",
       "2   [{'system': 'GDC', 'value': 'b936c96e-70ab-46e...   \n",
       "3   [{'system': 'GDC', 'value': 'be68cfb0-c47f-49a...   \n",
       "4   [{'system': 'GDC', 'value': '99b34e59-a5d0-4a8...   \n",
       "..                                                ...   \n",
       "95  [{'system': 'GDC', 'value': '6acf365b-19ce-4c9...   \n",
       "96  [{'system': 'GDC', 'value': '94f90e9a-ecfe-468...   \n",
       "97  [{'system': 'GDC', 'value': 'e1bd6dfa-0567-46e...   \n",
       "98  [{'system': 'GDC', 'value': '142a1357-f1e8-40b...   \n",
       "99  [{'system': 'GDC', 'value': 'd86c3d4a-861b-421...   \n",
       "\n",
       "   member_of_research_project primary_diagnosis_condition  \\\n",
       "0                   GENIE-NKI              Neoplasms, NOS   \n",
       "1                   GENIE-JHU              Neoplasms, NOS   \n",
       "2                   GENIE-NKI              Neoplasms, NOS   \n",
       "3                   GENIE-JHU              Neoplasms, NOS   \n",
       "4                   GENIE-JHU              Neoplasms, NOS   \n",
       "..                        ...                         ...   \n",
       "95                    CPTAC-3                     Gliomas   \n",
       "96                    CPTAC-3                     Gliomas   \n",
       "97                    CPTAC-3                     Gliomas   \n",
       "98                    CPTAC-3                     Gliomas   \n",
       "99                    CPTAC-3                     Gliomas   \n",
       "\n",
       "   primary_diagnosis_site       subject_id  \n",
       "0                   Brain   GENIE-NKI-8IN3  \n",
       "1                   Brain  GENIE-JHU-01660  \n",
       "2                   Brain   GENIE-NKI-U4QV  \n",
       "3                   Brain  GENIE-JHU-02615  \n",
       "4                   Brain  GENIE-JHU-02402  \n",
       "..                    ...              ...  \n",
       "95                  Brain        C3L-02642  \n",
       "96                  Brain        C3N-01334  \n",
       "97                  Brain        C3L-00674  \n",
       "98                  Brain        C3N-01852  \n",
       "99                  Brain        C3L-03260  \n",
       "\n",
       "[100 rows x 6 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "researchsubjectresults.to_dataframe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4564beee",
   "metadata": {},
   "source": [
    "Each row from the researchsubject endpoint results tells us about a subject in a given study. Using this endpoint we can find out information like what studies fit our search criteria, and also get data that we can filter to have only subjects from multiple studies, or only subjects from single studies.\n",
    "\n",
    "Any given subject will have one row per study they participated in. The subject_id in the last column of this view is the same as the `id` in the first column of the Subjects endpoint results. You can use this to combine information across endpoints, which is covered near the end of the <a href=\"../BuildingACohort/#merging-results-across-endpoints\">Cohort Building workflow</a> notebook.\n",
    "\n",
    "---\n",
    "\n",
    "<div class=\"cdadefine\" style=\"background-color:#add9e5;color:black;padding:20px;\">\n",
    "\n",
    "<h3>ResearchSubject Field Definitions</h3>\n",
    "\n",
    "<i>A research subject is the entity of interest in a specific research study or project, typically a human being or an animal, but can also be a device, group of humans or animals, or a tissue sample. Human research subjects are usually not traceable to a particular person to protect the subjects privacy. This entity plays the role of the case_id in existing data. An individual who participates in 3 studies will have 3 researchsubject IDs</i>\n",
    "    \n",
    "<ul>\n",
    "<li><b>id:</b> The 'logical' identifier of the entity in the system of record, e.g. a UUID. This 'id' is unique within a given system. The identified entity may have a different 'id' in a different system. For CDA, this is case_id.</li>\n",
    "<li><b>identifier:</b> A 'business' identifier for the entity, typically as provided by an external system or authority, that persists across implementing systems (i.e. a 'logical' identifier). Uses a specialized, complex 'Identifier' data type to capture information about the source of the business identifier - or a URI expressed as a string to an existing entity.</li>\n",
    "<li><b>identifier.system:</b> The system or namespace that defines the identifier.</li>\n",
    "<li><b>identifier.value:</b> The value of the identifier, as defined by the system.</li>\n",
    "<li><b>member_of_research_project:</b> A reference to the Study(s) of which this ResearchSubject is a member.</li>\n",
    "<li><b>primary_diagnosis_condition:</b> The text term used to describe the type of malignant disease, as categorized by the World Health Organization's (WHO) <a href=\"https://www.who.int/standards/classifications/other-classifications/international-classification-of-diseases-for-oncology\">International Classification of Diseases for Oncology</a> (ICD-O). This attribute represents the disease that qualified the subject for inclusion on the ResearchProject.</li>\n",
    "<li><b>primary_diagnosis_site:</b> The text term used to describe the primary site of disease, as categorized by the World Health Organization's (WHO) <a href=\"https://www.who.int/standards/classifications/other-classifications/international-classification-of-diseases-for-oncology\">International Classification of Diseases for Oncology</a> (ICD-O). This categorization groups cases into general categories. This attribute represents the primary site of disease that qualified the subject for inclusion on the ResearchProject.</li>\n",
    "<li><b>subject_id:</b> The 'logical' identifier of the entity in the system of record, e.g. a UUID. This 'id' is unique within a given system. The identified entity may have a different 'id' in a different system. Can be joined to the `id` field from subject results</li>\n",
    "</ul>  \n",
    "\n",
    "</div>\n",
    "    \n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47372646",
   "metadata": {},
   "source": [
    "### diagnosis\n",
    "\n",
    "The diagnosis endpoint is an extension of the researchsubject endpoint, and returns information about researchsubjects that have a diagnosis that meets our search criteria:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "2f80fa21",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3659 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>primary_diagnosis</th>\n",
       "      <th>age_at_diagnosis</th>\n",
       "      <th>morphology</th>\n",
       "      <th>stage</th>\n",
       "      <th>grade</th>\n",
       "      <th>method_of_diagnosis</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>researchsubject_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1e36c16c-2268-4c15-98a2-502e61ad4d84</td>\n",
       "      <td>[{'system': 'GDC', 'value': '1e36c16c-2268-4c1...</td>\n",
       "      <td>Neoplasm, malignant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8000/3</td>\n",
       "      <td>None</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>None</td>\n",
       "      <td>GENIE-GRCC-87432b67</td>\n",
       "      <td>1a25a33b-7d0d-4fa2-bff2-8c688bea40bb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>39b283b2-5abd-554d-8660-64d79e55455d</td>\n",
       "      <td>[{'system': 'GDC', 'value': '39b283b2-5abd-554...</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>21860.0</td>\n",
       "      <td>9440/3</td>\n",
       "      <td>None</td>\n",
       "      <td>not reported</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-32-2495</td>\n",
       "      <td>e7d728e5-1a9f-4f27-9604-272e1235e08a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3b3e9cd7-2848-11ec-b712-0a4e2186f121</td>\n",
       "      <td>[{'system': 'PDC', 'value': '3b3e9cd7-2848-11e...</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>16667.0</td>\n",
       "      <td>9440/3</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td></td>\n",
       "      <td>C3L-00674</td>\n",
       "      <td>104c0976-2139-11ea-aee1-0e1aae319e49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3dd9d358-2848-11ec-b712-0a4e2186f121</td>\n",
       "      <td>[{'system': 'PDC', 'value': '3dd9d358-2848-11e...</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>18038.0</td>\n",
       "      <td>9440/3</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td></td>\n",
       "      <td>C3N-00661</td>\n",
       "      <td>104c47ef-2139-11ea-aee1-0e1aae319e49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3e472906-2848-11ec-b712-0a4e2186f121</td>\n",
       "      <td>[{'system': 'PDC', 'value': '3e472906-2848-11e...</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>16195.0</td>\n",
       "      <td>9440/3</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td></td>\n",
       "      <td>C3N-01798</td>\n",
       "      <td>104c5b6d-2139-11ea-aee1-0e1aae319e49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>3ca82933-2848-11ec-b712-0a4e2186f121</td>\n",
       "      <td>[{'system': 'PDC', 'value': '3ca82933-2848-11e...</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>22025.0</td>\n",
       "      <td>9440/3</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td></td>\n",
       "      <td>C3L-02705</td>\n",
       "      <td>104c291d-2139-11ea-aee1-0e1aae319e49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>3ee64181-2848-11ec-b712-0a4e2186f121</td>\n",
       "      <td>[{'system': 'PDC', 'value': '3ee64181-2848-11e...</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>17506.0</td>\n",
       "      <td>9440/3</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td></td>\n",
       "      <td>C3N-02782</td>\n",
       "      <td>104c74a2-2139-11ea-aee1-0e1aae319e49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>4605156b-7d36-470b-a61c-d3bacbbccd63</td>\n",
       "      <td>[{'system': 'GDC', 'value': '4605156b-7d36-470...</td>\n",
       "      <td>Glioblastoma</td>\n",
       "      <td>24055.0</td>\n",
       "      <td>9440/3</td>\n",
       "      <td>None</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>None</td>\n",
       "      <td>C3N-00665</td>\n",
       "      <td>3d9ab180-73ab-4938-890f-cf345bfd4cf9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>4acbc5af-0a90-5b09-bd55-15dd2431c91c</td>\n",
       "      <td>[{'system': 'GDC', 'value': '4acbc5af-0a90-5b0...</td>\n",
       "      <td>Oligodendroglioma, NOS</td>\n",
       "      <td>14548.0</td>\n",
       "      <td>9450/3</td>\n",
       "      <td>None</td>\n",
       "      <td>not reported</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-HW-8322</td>\n",
       "      <td>6105c037-90f1-42d9-ac38-6e2252999a64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>66d7d6e6-b406-5ce8-b499-4d2840b532cc</td>\n",
       "      <td>[{'system': 'GDC', 'value': '66d7d6e6-b406-5ce...</td>\n",
       "      <td>Mixed glioma</td>\n",
       "      <td>15986.0</td>\n",
       "      <td>9382/3</td>\n",
       "      <td>None</td>\n",
       "      <td>not reported</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-DU-5872</td>\n",
       "      <td>48607130-c09a-491d-882f-7d2591d13182</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      id  \\\n",
       "0   1e36c16c-2268-4c15-98a2-502e61ad4d84   \n",
       "1   39b283b2-5abd-554d-8660-64d79e55455d   \n",
       "2   3b3e9cd7-2848-11ec-b712-0a4e2186f121   \n",
       "3   3dd9d358-2848-11ec-b712-0a4e2186f121   \n",
       "4   3e472906-2848-11ec-b712-0a4e2186f121   \n",
       "..                                   ...   \n",
       "95  3ca82933-2848-11ec-b712-0a4e2186f121   \n",
       "96  3ee64181-2848-11ec-b712-0a4e2186f121   \n",
       "97  4605156b-7d36-470b-a61c-d3bacbbccd63   \n",
       "98  4acbc5af-0a90-5b09-bd55-15dd2431c91c   \n",
       "99  66d7d6e6-b406-5ce8-b499-4d2840b532cc   \n",
       "\n",
       "                                           identifier       primary_diagnosis  \\\n",
       "0   [{'system': 'GDC', 'value': '1e36c16c-2268-4c1...     Neoplasm, malignant   \n",
       "1   [{'system': 'GDC', 'value': '39b283b2-5abd-554...            Glioblastoma   \n",
       "2   [{'system': 'PDC', 'value': '3b3e9cd7-2848-11e...            Glioblastoma   \n",
       "3   [{'system': 'PDC', 'value': '3dd9d358-2848-11e...            Glioblastoma   \n",
       "4   [{'system': 'PDC', 'value': '3e472906-2848-11e...            Glioblastoma   \n",
       "..                                                ...                     ...   \n",
       "95  [{'system': 'PDC', 'value': '3ca82933-2848-11e...            Glioblastoma   \n",
       "96  [{'system': 'PDC', 'value': '3ee64181-2848-11e...            Glioblastoma   \n",
       "97  [{'system': 'GDC', 'value': '4605156b-7d36-470...            Glioblastoma   \n",
       "98  [{'system': 'GDC', 'value': '4acbc5af-0a90-5b0...  Oligodendroglioma, NOS   \n",
       "99  [{'system': 'GDC', 'value': '66d7d6e6-b406-5ce...            Mixed glioma   \n",
       "\n",
       "    age_at_diagnosis morphology         stage         grade  \\\n",
       "0                NaN     8000/3          None  Not Reported   \n",
       "1            21860.0     9440/3          None  not reported   \n",
       "2            16667.0     9440/3  Not Reported  Not Reported   \n",
       "3            18038.0     9440/3  Not Reported  Not Reported   \n",
       "4            16195.0     9440/3  Not Reported  Not Reported   \n",
       "..               ...        ...           ...           ...   \n",
       "95           22025.0     9440/3  Not Reported  Not Reported   \n",
       "96           17506.0     9440/3  Not Reported  Not Reported   \n",
       "97           24055.0     9440/3          None  Not Reported   \n",
       "98           14548.0     9450/3          None  not reported   \n",
       "99           15986.0     9382/3          None  not reported   \n",
       "\n",
       "   method_of_diagnosis           subject_id  \\\n",
       "0                 None  GENIE-GRCC-87432b67   \n",
       "1                 None         TCGA-32-2495   \n",
       "2                                 C3L-00674   \n",
       "3                                 C3N-00661   \n",
       "4                                 C3N-01798   \n",
       "..                 ...                  ...   \n",
       "95                                C3L-02705   \n",
       "96                                C3N-02782   \n",
       "97                None            C3N-00665   \n",
       "98                None         TCGA-HW-8322   \n",
       "99                None         TCGA-DU-5872   \n",
       "\n",
       "                      researchsubject_id  \n",
       "0   1a25a33b-7d0d-4fa2-bff2-8c688bea40bb  \n",
       "1   e7d728e5-1a9f-4f27-9604-272e1235e08a  \n",
       "2   104c0976-2139-11ea-aee1-0e1aae319e49  \n",
       "3   104c47ef-2139-11ea-aee1-0e1aae319e49  \n",
       "4   104c5b6d-2139-11ea-aee1-0e1aae319e49  \n",
       "..                                   ...  \n",
       "95  104c291d-2139-11ea-aee1-0e1aae319e49  \n",
       "96  104c74a2-2139-11ea-aee1-0e1aae319e49  \n",
       "97  3d9ab180-73ab-4938-890f-cf345bfd4cf9  \n",
       "98  6105c037-90f1-42d9-ac38-6e2252999a64  \n",
       "99  48607130-c09a-491d-882f-7d2591d13182  \n",
       "\n",
       "[100 rows x 10 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "diagnosisresults = myquery.diagnosis.run()\n",
    "diagnosisresults.to_dataframe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6008331",
   "metadata": {},
   "source": [
    "\n",
    "---\n",
    "\n",
    "<div class=\"cdadefine\" style=\"background-color:#add9e5;color:black;padding:20px;\">\n",
    "\n",
    "<h3>Diagnosis Field Definitions</h3>\n",
    "\n",
    "<i>A collection of characteristics that describe an abnormal condition of the body as assessed at a point in time. May be used to capture information about neoplastic and non-neoplastic conditions.</i>\n",
    "\n",
    "    \n",
    "<ul>\n",
    "  <li><b>id:</b> The 'logical' identifier of the entity in the repository, e.g. a UUID. This 'id' is unique within a given system. The identified entity may have a different 'id' in a different system.</li>\n",
    "<li><b>identifier:</b> A 'business' identifier for the entity, typically as provided by an external system or authority, that persists across implementing systems (i.e. a 'logical' identifier). Uses a specialized, complex 'Identifier' data type to capture information about the source of the business identifier - or a URI expressed as a string to an existing entity.</li>\n",
    "<li><b>identifier.system:</b> The system or namespace that defines the identifier.</li>\n",
    "<li><b>identifier.value:</b> The value of the identifier, as defined by the system.</li>\n",
    "  <li><b>primary_diagnosis:</b> The diagnosis instance that qualified a subject for inclusion on a ResearchProject.</li>\n",
    "  <li><b>age_at_diagnosis:</b> The age in days of the individual at the time of diagnosis.</li>\n",
    "  <li><b>morphology:</b> Code that represents the histology of the disease using the third edition of the <a href=\"https://www.who.int/standards/classifications/other-classifications/international-classification-of-diseases-for-oncology\">International Classification of Diseases for Oncology</a>, published in 2000, used principally in tumor and cancer registries for coding the site (topography) and the histology (morphology) of neoplasms.</li>\n",
    "  <li><b>stage:</b> The extent of a cancer in the body. Staging is usually based on the size of the tumor, whether lymph nodes contain cancer, and whether the cancer has spread from the original site to other parts of the body. Different diseases may use different staging criteria, please refer to the originating data source to see what staging system is reported</li>\n",
    "  <li><b>grade:</b> The degree of abnormality of cancer cells, a measure of differentiation, the extent to which cancer cells are similar in appearance and function to healthy cells of the same tissue type. The degree of differentiation often relates to the clinical behavior of the particular tumor. Based on the microscopic findings, tumor grade is commonly described by one of four degrees of severity. Histopathologic grade of a tumor may be used to plan treatment and estimate the future course, outcome, and overall prognosis of disease. Certain types of cancers, such as soft tissue sarcoma, primary brain tumors, lymphomas, and breast have special grading systems.</li>\n",
    "  <li><b>method_of_diagnosis:</b> The method used to confirm the patients malignant diagnosis.</li>\n",
    "  <li><b>subject_id:</b> An identifier for the subject. Can be joined to the `id` field from subject results</li>\n",
    "  <li><b>researchsubject_id:</b> An identifier for the subject. Can be joined to the `id` field from researchsubject results</li>\n",
    "</ul>  \n",
    "\n",
    "</div>\n",
    "    \n",
    "---\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5855d4dd",
   "metadata": {},
   "source": [
    "### treatment\n",
    "\n",
    "The treatment endpoint is an extension of diagnosis and returns information about treatments undertaken on research subjects that have a given diagnosis that meets our search criteria:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d81bdba0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3533 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>treatment_type</th>\n",
       "      <th>treatment_outcome</th>\n",
       "      <th>days_to_treatment_start</th>\n",
       "      <th>days_to_treatment_end</th>\n",
       "      <th>therapeutic_agent</th>\n",
       "      <th>treatment_anatomic_site</th>\n",
       "      <th>treatment_effect</th>\n",
       "      <th>treatment_end_reason</th>\n",
       "      <th>number_of_cycles</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>researchsubject_id</th>\n",
       "      <th>researchsubject_diagnosis_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>035be2c6-6d94-5fd3-bb3c-070440d2d5b9</td>\n",
       "      <td>[{'system': 'GDC', 'value': '035be2c6-6d94-5fd...</td>\n",
       "      <td>Pharmaceutical Therapy, NOS</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-DU-8161</td>\n",
       "      <td>b9a3b647-61b3-4d98-90ca-e307ab50984a</td>\n",
       "      <td>5662ed5f-1e32-5699-a005-0edf7429bd3a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>04215940-62d7-5fba-bed4-32728b696c40</td>\n",
       "      <td>[{'system': 'GDC', 'value': '04215940-62d7-5fb...</td>\n",
       "      <td>Radiation Therapy, NOS</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-19-0960</td>\n",
       "      <td>37a47ea2-1daa-44a4-ade5-84d383c2b13c</td>\n",
       "      <td>12bc37c7-6674-5154-8dae-f0a4e5e0096e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0dc4e132-d0ca-55d0-b6f6-360af92dd2a8</td>\n",
       "      <td>[{'system': 'GDC', 'value': '0dc4e132-d0ca-55d...</td>\n",
       "      <td>Pharmaceutical Therapy, NOS</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-FG-8182</td>\n",
       "      <td>6a0bcf0c-fa4c-4119-99d2-f722b781d20f</td>\n",
       "      <td>d5301d3a-b057-5cc2-a370-5aadad7a9b55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>17b6d316-b324-5539-8067-c2419cd88ac8</td>\n",
       "      <td>[{'system': 'GDC', 'value': '17b6d316-b324-553...</td>\n",
       "      <td>Pharmaceutical Therapy, NOS</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-DU-5855</td>\n",
       "      <td>e326c1dd-2ef8-431b-a657-440aa3f35ff3</td>\n",
       "      <td>2787e928-f6e7-5053-a07a-bb1a696390fb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1cc2099c-0bec-5ba9-9715-438a6ad29950</td>\n",
       "      <td>[{'system': 'GDC', 'value': '1cc2099c-0bec-5ba...</td>\n",
       "      <td>Radiation Therapy, NOS</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-S9-A6U9</td>\n",
       "      <td>38eca720-ad68-43a7-8a1c-f79f838d5d03</td>\n",
       "      <td>f3cd7529-db2f-56f3-bb06-0cdaabf1a775</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>b806fc69-ec07-5b26-8b96-dd0573dab74f</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'b806fc69-ec07-5b2...</td>\n",
       "      <td>Pharmaceutical Therapy, NOS</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-S9-A7J3</td>\n",
       "      <td>032096a6-1923-44b9-a1cb-181bed041c8e</td>\n",
       "      <td>fc35b9d4-a743-5bca-ab74-0aa81f5b6369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>ba38b0d3-9eab-59d8-a94c-257a5ff8d345</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'ba38b0d3-9eab-59d...</td>\n",
       "      <td>Pharmaceutical Therapy, NOS</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-19-0963</td>\n",
       "      <td>163e93a7-4d8d-4f28-a9f6-4d0ed16630cd</td>\n",
       "      <td>45ae8847-a98e-5bea-a7b8-31f78c5e09a8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>ce366c44-ae99-590f-af64-c98a5b17d13b</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'ce366c44-ae99-590...</td>\n",
       "      <td>Pharmaceutical Therapy, NOS</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-08-0389</td>\n",
       "      <td>ea980264-d64d-410a-86dd-c129d7d8d1ee</td>\n",
       "      <td>96944e9a-22f6-5ac4-9ae7-1cd512297d3e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>ce369bc9-a7ce-5b9e-a1cd-cba77c86dbc6</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'ce369bc9-a7ce-5b9...</td>\n",
       "      <td>Radiation Therapy, NOS</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-S9-A7R4</td>\n",
       "      <td>58650d08-7685-41de-9ae0-21665f8f449f</td>\n",
       "      <td>bcac39f9-4ba0-599e-85b8-a529e361d038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>d3f5a531-2e70-5f84-b6cd-6186406abd36</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'd3f5a531-2e70-5f8...</td>\n",
       "      <td>Pharmaceutical Therapy, NOS</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>TCGA-FG-5963</td>\n",
       "      <td>68f550ca-3a33-45e4-b1c9-14f99e543263</td>\n",
       "      <td>df157cdd-174a-539c-ae8f-698a3b3a46a2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      id  \\\n",
       "0   035be2c6-6d94-5fd3-bb3c-070440d2d5b9   \n",
       "1   04215940-62d7-5fba-bed4-32728b696c40   \n",
       "2   0dc4e132-d0ca-55d0-b6f6-360af92dd2a8   \n",
       "3   17b6d316-b324-5539-8067-c2419cd88ac8   \n",
       "4   1cc2099c-0bec-5ba9-9715-438a6ad29950   \n",
       "..                                   ...   \n",
       "95  b806fc69-ec07-5b26-8b96-dd0573dab74f   \n",
       "96  ba38b0d3-9eab-59d8-a94c-257a5ff8d345   \n",
       "97  ce366c44-ae99-590f-af64-c98a5b17d13b   \n",
       "98  ce369bc9-a7ce-5b9e-a1cd-cba77c86dbc6   \n",
       "99  d3f5a531-2e70-5f84-b6cd-6186406abd36   \n",
       "\n",
       "                                           identifier  \\\n",
       "0   [{'system': 'GDC', 'value': '035be2c6-6d94-5fd...   \n",
       "1   [{'system': 'GDC', 'value': '04215940-62d7-5fb...   \n",
       "2   [{'system': 'GDC', 'value': '0dc4e132-d0ca-55d...   \n",
       "3   [{'system': 'GDC', 'value': '17b6d316-b324-553...   \n",
       "4   [{'system': 'GDC', 'value': '1cc2099c-0bec-5ba...   \n",
       "..                                                ...   \n",
       "95  [{'system': 'GDC', 'value': 'b806fc69-ec07-5b2...   \n",
       "96  [{'system': 'GDC', 'value': 'ba38b0d3-9eab-59d...   \n",
       "97  [{'system': 'GDC', 'value': 'ce366c44-ae99-590...   \n",
       "98  [{'system': 'GDC', 'value': 'ce369bc9-a7ce-5b9...   \n",
       "99  [{'system': 'GDC', 'value': 'd3f5a531-2e70-5f8...   \n",
       "\n",
       "                 treatment_type treatment_outcome  days_to_treatment_start  \\\n",
       "0   Pharmaceutical Therapy, NOS              None                      NaN   \n",
       "1        Radiation Therapy, NOS              None                      NaN   \n",
       "2   Pharmaceutical Therapy, NOS              None                      NaN   \n",
       "3   Pharmaceutical Therapy, NOS              None                      NaN   \n",
       "4        Radiation Therapy, NOS              None                      NaN   \n",
       "..                          ...               ...                      ...   \n",
       "95  Pharmaceutical Therapy, NOS              None                      NaN   \n",
       "96  Pharmaceutical Therapy, NOS              None                      NaN   \n",
       "97  Pharmaceutical Therapy, NOS              None                      NaN   \n",
       "98       Radiation Therapy, NOS              None                      NaN   \n",
       "99  Pharmaceutical Therapy, NOS              None                      NaN   \n",
       "\n",
       "   days_to_treatment_end therapeutic_agent treatment_anatomic_site  \\\n",
       "0                   None              None                    None   \n",
       "1                   None              None                    None   \n",
       "2                   None              None                    None   \n",
       "3                   None              None                    None   \n",
       "4                   None              None                    None   \n",
       "..                   ...               ...                     ...   \n",
       "95                  None              None                    None   \n",
       "96                  None              None                    None   \n",
       "97                  None              None                    None   \n",
       "98                  None              None                    None   \n",
       "99                  None              None                    None   \n",
       "\n",
       "   treatment_effect treatment_end_reason number_of_cycles    subject_id  \\\n",
       "0              None                 None             None  TCGA-DU-8161   \n",
       "1              None                 None             None  TCGA-19-0960   \n",
       "2              None                 None             None  TCGA-FG-8182   \n",
       "3              None                 None             None  TCGA-DU-5855   \n",
       "4              None                 None             None  TCGA-S9-A6U9   \n",
       "..              ...                  ...              ...           ...   \n",
       "95             None                 None             None  TCGA-S9-A7J3   \n",
       "96             None                 None             None  TCGA-19-0963   \n",
       "97             None                 None             None  TCGA-08-0389   \n",
       "98             None                 None             None  TCGA-S9-A7R4   \n",
       "99             None                 None             None  TCGA-FG-5963   \n",
       "\n",
       "                      researchsubject_id          researchsubject_diagnosis_id  \n",
       "0   b9a3b647-61b3-4d98-90ca-e307ab50984a  5662ed5f-1e32-5699-a005-0edf7429bd3a  \n",
       "1   37a47ea2-1daa-44a4-ade5-84d383c2b13c  12bc37c7-6674-5154-8dae-f0a4e5e0096e  \n",
       "2   6a0bcf0c-fa4c-4119-99d2-f722b781d20f  d5301d3a-b057-5cc2-a370-5aadad7a9b55  \n",
       "3   e326c1dd-2ef8-431b-a657-440aa3f35ff3  2787e928-f6e7-5053-a07a-bb1a696390fb  \n",
       "4   38eca720-ad68-43a7-8a1c-f79f838d5d03  f3cd7529-db2f-56f3-bb06-0cdaabf1a775  \n",
       "..                                   ...                                   ...  \n",
       "95  032096a6-1923-44b9-a1cb-181bed041c8e  fc35b9d4-a743-5bca-ab74-0aa81f5b6369  \n",
       "96  163e93a7-4d8d-4f28-a9f6-4d0ed16630cd  45ae8847-a98e-5bea-a7b8-31f78c5e09a8  \n",
       "97  ea980264-d64d-410a-86dd-c129d7d8d1ee  96944e9a-22f6-5ac4-9ae7-1cd512297d3e  \n",
       "98  58650d08-7685-41de-9ae0-21665f8f449f  bcac39f9-4ba0-599e-85b8-a529e361d038  \n",
       "99  68f550ca-3a33-45e4-b1c9-14f99e543263  df157cdd-174a-539c-ae8f-698a3b3a46a2  \n",
       "\n",
       "[100 rows x 14 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "treatmentresults = myquery.treatment.run()\n",
    "treatmentresults.to_dataframe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30804682",
   "metadata": {},
   "source": [
    "\n",
    "---\n",
    "\n",
    "<div class=\"cdadefine\" style=\"background-color:#add9e5;color:black;padding:20px;\">\n",
    "\n",
    "<h3>Treatment Field Definitions</h3>\n",
    "\n",
    "<i><i> Medication administration or other treatment types. A single research subject may have multiple treatments for a single diagnosis, and/or different diagnoses, and different treatments, across different studies</i></i>\n",
    "\n",
    "    \n",
    "<ul>\n",
    "  <li><b>id:</b> The 'logical' identifier of the entity in the repository, e.g. a UUID. This 'id' is unique within a given system. The identified entity may have a different 'id' in a different system.</li>\n",
    "<li><b>identifier:</b> A 'business' identifier for the entity, typically as provided by an external system or authority, that persists across implementing systems (i.e. a 'logical' identifier). Uses a specialized, complex 'Identifier' data type to capture information about the source of the business identifier - or a URI expressed as a string to an existing entity.</li>\n",
    "<li><b>identifier.system:</b> The system or namespace that defines the identifier.</li>\n",
    "<li><b>identifier.value:</b> The value of the identifier, as defined by the system.</li>\n",
    "  <li><b>treatment_type:</b> The treatment type including medication/therapeutics or other procedures.</li>\n",
    "  <li><b>treatment_outcome:</b> The final outcome of the treatment.</li>\n",
    "  <li><b>days_to_treatment_start:</b> The timepoint at which the treatment started.</li>\n",
    "  <li><b>days_to_treatment_end:</b>The timepoint at which the treatment ended. </li>\n",
    "  <li><b>therapeutic_agent:</b> One or more therapeutic agents as part of this treatment.</li>\n",
    "  <li><b>treatment_anatomic_site:</b> The anatomical site that the treatment targets.</li>\n",
    "  <li><b>treatment_effect:</b>The effect of a treatment on the diagnosis or tumor. </li>\n",
    "  <li><b>treatment_end_reason:</b>The reason the treatment ended. </li>\n",
    "  <li><b>number_of_cycles:</b>The number of treatment cycles the subject received. </li>\n",
    "  <li><b>subject_id:</b> An identifier for the subject. Can be joined to the `id` field from subject results</li>\n",
    "  <li><b>researchsubject_id:</b> An identifier for the researchsubject. Can be joined to the `id` field from researchsubject results</li>\n",
    "  <li><b>researchsubject_diagnosis_id:</b> An identifier for the diagnosis. Can be joined to the `id` field from diagnosis results</li>\n",
    "</ul>  \n",
    "\n",
    "</div>\n",
    "    \n",
    "---\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c00bec8",
   "metadata": {},
   "source": [
    "### specimen\n",
    "\n",
    "We can use this same query to see what specimens are available for brain tissue at the CDA:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "51960eed",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 4069 ms\n",
      "\n",
      "            QueryID: fdf0c44f-5d00-4917-83f4-282484d4c971\n",
      "            \n",
      "            Offset: 0\n",
      "            Count: 100\n",
      "            Total Row Count: 39201\n",
      "            More pages: True\n",
      "            \n"
     ]
    }
   ],
   "source": [
    "specimenresults =  myquery.specimen.run()\n",
    "print(specimenresults)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f254547",
   "metadata": {},
   "source": [
    "Nearly 40,000 specimens meet our search criteria! We would typically expect this number to be much larger than our number of subjects or researchsubjects. First because studies will often take more than one sample per subject, and second because any given specimen might be aliquoted out to be used in multiple tests. Since we didn't specify any further filters, our results will return all of these as seperate speciments. Let's look at a few:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c55b0128",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>associated_project</th>\n",
       "      <th>days_to_collection</th>\n",
       "      <th>primary_disease_type</th>\n",
       "      <th>anatomical_site</th>\n",
       "      <th>source_material_type</th>\n",
       "      <th>specimen_type</th>\n",
       "      <th>derived_from_specimen</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>researchsubject_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0072c6f0-c876-4296-aabe-376ef49efab0</td>\n",
       "      <td>[{'system': 'GDC', 'value': '0072c6f0-c876-429...</td>\n",
       "      <td>TCGA-GBM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>None</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>analyte</td>\n",
       "      <td>f2b1a88f-baf0-4ddc-a527-b0fdc08c5145</td>\n",
       "      <td>TCGA-02-0046</td>\n",
       "      <td>5b8f05fa-0145-4142-8d41-5980e4d36d81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>01e4984a-e570-4623-b020-2ab576af9727</td>\n",
       "      <td>[{'system': 'GDC', 'value': '01e4984a-e570-462...</td>\n",
       "      <td>TCGA-LGG</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>None</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>slide</td>\n",
       "      <td>4f41877a-1b43-52c0-81cd-c7de3fc7e1da</td>\n",
       "      <td>TCGA-S9-A6WI</td>\n",
       "      <td>0b283f95-dbd4-4c36-9201-2198a90ee0d6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>025bebe3-cf8e-4c92-92b4-39bd4003f6ab</td>\n",
       "      <td>[{'system': 'GDC', 'value': '025bebe3-cf8e-4c9...</td>\n",
       "      <td>TCGA-LGG</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>None</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>slide</td>\n",
       "      <td>7dc77a07-bbc3-5056-9e39-016379e326a2</td>\n",
       "      <td>TCGA-R8-A6MK</td>\n",
       "      <td>65836b59-b469-4b89-adf5-e65735e494f5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>029fdf74-3d3a-48a6-b2b0-76d10fe7fd59</td>\n",
       "      <td>[{'system': 'GDC', 'value': '029fdf74-3d3a-48a...</td>\n",
       "      <td>TCGA-LGG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>None</td>\n",
       "      <td>Blood Derived Normal</td>\n",
       "      <td>aliquot</td>\n",
       "      <td>94ee050b-69be-422a-bc95-cec676e445f8</td>\n",
       "      <td>TCGA-HT-8105</td>\n",
       "      <td>d766c3af-f0b3-4791-a754-d6546d370f8f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>032dd99e-66bf-4b29-b9e9-952289f28d6a</td>\n",
       "      <td>[{'system': 'GDC', 'value': '032dd99e-66bf-4b2...</td>\n",
       "      <td>HCMI-CMDC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>None</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>portion</td>\n",
       "      <td>be4394ef-8a86-4eec-8b76-760773a23253</td>\n",
       "      <td>HCM-BROD-0613-C71</td>\n",
       "      <td>b3cc0754-6a21-4683-a491-b2fb8b6175a4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>3d4a149d-0b92-402c-af45-94675bee4710</td>\n",
       "      <td>[{'system': 'GDC', 'value': '3d4a149d-0b92-402...</td>\n",
       "      <td>TCGA-GBM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>None</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>aliquot</td>\n",
       "      <td>9d7d0534-0230-4af2-8df4-2adda1966077</td>\n",
       "      <td>TCGA-14-1451</td>\n",
       "      <td>06c427a1-8c2d-4003-b744-139114a04440</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>3df80cf4-56dd-48e0-9112-b0d730bdb2e9</td>\n",
       "      <td>[{'system': 'GDC', 'value': '3df80cf4-56dd-48e...</td>\n",
       "      <td>TCGA-GBM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>None</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>aliquot</td>\n",
       "      <td>47959142-1c3d-4764-8dbc-bcd91612f47d</td>\n",
       "      <td>TCGA-27-2528</td>\n",
       "      <td>be3a7ef3-34ed-40e1-9d9c-187940596b26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>3ec82364-e5f1-41d3-a4ec-9e81f7616165</td>\n",
       "      <td>[{'system': 'GDC', 'value': '3ec82364-e5f1-41d...</td>\n",
       "      <td>TCGA-GBM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>None</td>\n",
       "      <td>Solid Tissue Normal</td>\n",
       "      <td>analyte</td>\n",
       "      <td>863a25cc-8d7e-440e-b98d-72941e53a313</td>\n",
       "      <td>TCGA-08-0359</td>\n",
       "      <td>0aa5b68f-2b99-44d2-b17f-f4eb17fe891e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>3f20d55d-f1ad-4ede-9905-b456d7db4898</td>\n",
       "      <td>[{'system': 'GDC', 'value': '3f20d55d-f1ad-4ed...</td>\n",
       "      <td>TCGA-LGG</td>\n",
       "      <td>1243.0</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>None</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>slide</td>\n",
       "      <td>1e57b685-0cd6-421a-8e8e-c2f3820642f4</td>\n",
       "      <td>TCGA-TM-A84G</td>\n",
       "      <td>63f4aa26-dec0-4309-8b75-6446a7241527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>3f654f21-3b06-4562-b3a8-91577c6dc811</td>\n",
       "      <td>[{'system': 'GDC', 'value': '3f654f21-3b06-456...</td>\n",
       "      <td>TCGA-GBM</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Gliomas</td>\n",
       "      <td>None</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>slide</td>\n",
       "      <td>cee9cbc3-a3ab-58c5-ba8d-da38dfa69a78</td>\n",
       "      <td>TCGA-02-0034</td>\n",
       "      <td>69205453-c504-4684-8d32-57c82efc679d</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      id  \\\n",
       "0   0072c6f0-c876-4296-aabe-376ef49efab0   \n",
       "1   01e4984a-e570-4623-b020-2ab576af9727   \n",
       "2   025bebe3-cf8e-4c92-92b4-39bd4003f6ab   \n",
       "3   029fdf74-3d3a-48a6-b2b0-76d10fe7fd59   \n",
       "4   032dd99e-66bf-4b29-b9e9-952289f28d6a   \n",
       "..                                   ...   \n",
       "95  3d4a149d-0b92-402c-af45-94675bee4710   \n",
       "96  3df80cf4-56dd-48e0-9112-b0d730bdb2e9   \n",
       "97  3ec82364-e5f1-41d3-a4ec-9e81f7616165   \n",
       "98  3f20d55d-f1ad-4ede-9905-b456d7db4898   \n",
       "99  3f654f21-3b06-4562-b3a8-91577c6dc811   \n",
       "\n",
       "                                           identifier associated_project  \\\n",
       "0   [{'system': 'GDC', 'value': '0072c6f0-c876-429...           TCGA-GBM   \n",
       "1   [{'system': 'GDC', 'value': '01e4984a-e570-462...           TCGA-LGG   \n",
       "2   [{'system': 'GDC', 'value': '025bebe3-cf8e-4c9...           TCGA-LGG   \n",
       "3   [{'system': 'GDC', 'value': '029fdf74-3d3a-48a...           TCGA-LGG   \n",
       "4   [{'system': 'GDC', 'value': '032dd99e-66bf-4b2...          HCMI-CMDC   \n",
       "..                                                ...                ...   \n",
       "95  [{'system': 'GDC', 'value': '3d4a149d-0b92-402...           TCGA-GBM   \n",
       "96  [{'system': 'GDC', 'value': '3df80cf4-56dd-48e...           TCGA-GBM   \n",
       "97  [{'system': 'GDC', 'value': '3ec82364-e5f1-41d...           TCGA-GBM   \n",
       "98  [{'system': 'GDC', 'value': '3f20d55d-f1ad-4ed...           TCGA-LGG   \n",
       "99  [{'system': 'GDC', 'value': '3f654f21-3b06-456...           TCGA-GBM   \n",
       "\n",
       "    days_to_collection primary_disease_type anatomical_site  \\\n",
       "0                  NaN              Gliomas            None   \n",
       "1                  0.0              Gliomas            None   \n",
       "2                  0.0              Gliomas            None   \n",
       "3                  NaN              Gliomas            None   \n",
       "4                  NaN              Gliomas            None   \n",
       "..                 ...                  ...             ...   \n",
       "95                 NaN              Gliomas            None   \n",
       "96                 NaN              Gliomas            None   \n",
       "97                 NaN              Gliomas            None   \n",
       "98              1243.0              Gliomas            None   \n",
       "99                 0.0              Gliomas            None   \n",
       "\n",
       "    source_material_type specimen_type                 derived_from_specimen  \\\n",
       "0          Primary Tumor       analyte  f2b1a88f-baf0-4ddc-a527-b0fdc08c5145   \n",
       "1          Primary Tumor         slide  4f41877a-1b43-52c0-81cd-c7de3fc7e1da   \n",
       "2          Primary Tumor         slide  7dc77a07-bbc3-5056-9e39-016379e326a2   \n",
       "3   Blood Derived Normal       aliquot  94ee050b-69be-422a-bc95-cec676e445f8   \n",
       "4          Primary Tumor       portion  be4394ef-8a86-4eec-8b76-760773a23253   \n",
       "..                   ...           ...                                   ...   \n",
       "95         Primary Tumor       aliquot  9d7d0534-0230-4af2-8df4-2adda1966077   \n",
       "96         Primary Tumor       aliquot  47959142-1c3d-4764-8dbc-bcd91612f47d   \n",
       "97   Solid Tissue Normal       analyte  863a25cc-8d7e-440e-b98d-72941e53a313   \n",
       "98         Primary Tumor         slide  1e57b685-0cd6-421a-8e8e-c2f3820642f4   \n",
       "99         Primary Tumor         slide  cee9cbc3-a3ab-58c5-ba8d-da38dfa69a78   \n",
       "\n",
       "           subject_id                    researchsubject_id  \n",
       "0        TCGA-02-0046  5b8f05fa-0145-4142-8d41-5980e4d36d81  \n",
       "1        TCGA-S9-A6WI  0b283f95-dbd4-4c36-9201-2198a90ee0d6  \n",
       "2        TCGA-R8-A6MK  65836b59-b469-4b89-adf5-e65735e494f5  \n",
       "3        TCGA-HT-8105  d766c3af-f0b3-4791-a754-d6546d370f8f  \n",
       "4   HCM-BROD-0613-C71  b3cc0754-6a21-4683-a491-b2fb8b6175a4  \n",
       "..                ...                                   ...  \n",
       "95       TCGA-14-1451  06c427a1-8c2d-4003-b744-139114a04440  \n",
       "96       TCGA-27-2528  be3a7ef3-34ed-40e1-9d9c-187940596b26  \n",
       "97       TCGA-08-0359  0aa5b68f-2b99-44d2-b17f-f4eb17fe891e  \n",
       "98       TCGA-TM-A84G  63f4aa26-dec0-4309-8b75-6446a7241527  \n",
       "99       TCGA-02-0034  69205453-c504-4684-8d32-57c82efc679d  \n",
       "\n",
       "[100 rows x 11 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "specimenresults.to_dataframe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ccb0f1c",
   "metadata": {},
   "source": [
    "\n",
    "---\n",
    "\n",
    "<div class=\"cdadefine\" style=\"background-color:#add9e5;color:black;padding:20px;\">\n",
    "\n",
    "<h3>Specimen Field Definitions</h3>\n",
    "\n",
    "<i><i>Any material taken as a sample from a biological entity (living or dead), or from a physical object or the environment. Specimens are usually collected as an example of their kind, often for use in some investigation.</i>\n",
    " A given specimen will have only a single subject ID and a single research subject ID</i>\n",
    "    \n",
    "<ul>\n",
    "  <li><b>id:</b> The 'logical' identifier of the entity in the repository, e.g. a UUID. This 'id' is unique within a given system. The identified entity may have a different 'id' in a different system.</li>\n",
    "<li><b>identifier:</b> A 'business' identifier for the entity, typically as provided by an external system or authority, that persists across implementing systems (i.e. a 'logical' identifier). Uses a specialized, complex 'Identifier' data type to capture information about the source of the business identifier - or a URI expressed as a string to an existing entity.</li>\n",
    "<li><b>identifier.system:</b> The system or namespace that defines the identifier.</li>\n",
    "<li><b>identifier.value:</b> The value of the identifier, as defined by the system.</li>\n",
    "  <li><b>associated_project:</b> The Project associated with the specimen.</li>\n",
    "  <li><b>days_to_collection:</b> The number of days from the index date to either the date a sample was collected for a specific study or project, or the date a patient underwent a procedure (e.g. surgical resection) yielding a sample that was eventually used for research.</li>\n",
    "  <li><b>primary_disease_type:</b> The text term used to describe the type of malignant disease, as categorized by the World Health Organization's (WHO) International Classification of Diseases for Oncology (ICD-O). This attribute represents the disease that qualified the subject for inclusion on the ResearchProject.</li>\n",
    "  <li><b>anatomical_site:</b> Per GDC Dictionary, the text term that represents the name of the primary disease site of the submitted tumor sample; recommend dropping tumor; biospecimen_anatomic_site.</li>\n",
    "  <li><b>source_material_type:</b> The general kind of material from which the specimen was derived, indicating the physical nature of the source material.</li>\n",
    "  <li><b>specimen_type:</b> The high-level type of the specimen, based on its how it has been derived from the original extracted sample. One of: analyte, aliquot, portion, sample, or slide.</li>\n",
    "  <li><b>derived_from_specimen:</b> A source/parent specimen from which this one was directly derived.</li>\n",
    "  <li><b>subject_id:</b> An identifier for the subject. Can be joined to the `id` field from subject results</li>\n",
    "  <li><b>researchsubject_id:</b> An identifier for the subject. Can be joined to the `id` field from researchsubject results</li>\n",
    "</ul>  \n",
    "\n",
    "</div>\n",
    "    \n",
    "---\n",
    "\n",
    "\n",
    "### file\n",
    "\n",
    "The file endpoint returns information about files that meet our search criteria, regardless of whether they are attached to subjects, research-subjects or specimens: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "94a0361f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3858 ms\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "            QueryID: 10be2d40-2d6e-4b23-90ee-a67353c088ee\n",
       "            \n",
       "            Offset: 0\n",
       "            Count: 100\n",
       "            Total Row Count: 4138879\n",
       "            More pages: True\n",
       "            "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "myquery.file.run()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "6006ec89",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3764 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>label</th>\n",
       "      <th>data_category</th>\n",
       "      <th>data_type</th>\n",
       "      <th>file_format</th>\n",
       "      <th>associated_project</th>\n",
       "      <th>drs_uri</th>\n",
       "      <th>byte_size</th>\n",
       "      <th>checksum</th>\n",
       "      <th>data_modality</th>\n",
       "      <th>imaging_modality</th>\n",
       "      <th>dbgap_accession_number</th>\n",
       "      <th>imaging_series</th>\n",
       "      <th>researchsubject_specimen_id</th>\n",
       "      <th>researchsubject_id</th>\n",
       "      <th>subject_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>e5b8f4c3-efa1-412c-8769-6bd629ceff25</td>\n",
       "      <td>[{'system': 'PDC', 'value': 'e5b8f4c3-efa1-412...</td>\n",
       "      <td>09CPTAC_GBM_W_PNNL_20190501_B3S1_f24.mzid.gz</td>\n",
       "      <td>Peptide Spectral Matches</td>\n",
       "      <td>Open Standard</td>\n",
       "      <td>mzIdentML</td>\n",
       "      <td>CPTAC3-Discovery</td>\n",
       "      <td>drs://dg.4DFC:e5b8f4c3-efa1-412c-8769-6bd629ce...</td>\n",
       "      <td>8445510</td>\n",
       "      <td>b707f6f6abb880553146af8214911d06</td>\n",
       "      <td>Proteomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>C3N-01366</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>416a6c7e-15e5-11ea-9bfa-0a42f3c845fe</td>\n",
       "      <td>[{'system': 'PDC', 'value': '416a6c7e-15e5-11e...</td>\n",
       "      <td>06CPTAC_GBM_P_PNNL_20190306_B2S2_f12.raw</td>\n",
       "      <td>Raw Mass Spectra</td>\n",
       "      <td>Proprietary</td>\n",
       "      <td>vendor-specific</td>\n",
       "      <td>CPTAC3-Discovery</td>\n",
       "      <td>drs://dg.4DFC:416a6c7e-15e5-11ea-9bfa-0a42f3c8...</td>\n",
       "      <td>615428786</td>\n",
       "      <td>e2620190ade98fff9bffd0149feefd8d</td>\n",
       "      <td>Proteomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>C3N-02185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>32a4267d-0cc1-4022-ae45-5efb4a29eef6</td>\n",
       "      <td>[{'system': 'PDC', 'value': '32a4267d-0cc1-402...</td>\n",
       "      <td>08CPTAC_GBM_W_PNNL_20190306_B2S4_f18.mzML.gz</td>\n",
       "      <td>Processed Mass Spectra</td>\n",
       "      <td>Open Standard</td>\n",
       "      <td>mzML</td>\n",
       "      <td>CPTAC3-Discovery</td>\n",
       "      <td>drs://dg.4DFC:32a4267d-0cc1-4022-ae45-5efb4a29...</td>\n",
       "      <td>200563993</td>\n",
       "      <td>170fca4126b7a4977c5d2248f0b9b537</td>\n",
       "      <td>Proteomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>C3N-01856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6301f865-15e6-11ea-9bfa-0a42f3c845fe</td>\n",
       "      <td>[{'system': 'PDC', 'value': '6301f865-15e6-11e...</td>\n",
       "      <td>11CPTAC_GBM_W_PNNL_20190501_B3S3_f09.raw</td>\n",
       "      <td>Raw Mass Spectra</td>\n",
       "      <td>Proprietary</td>\n",
       "      <td>vendor-specific</td>\n",
       "      <td>CPTAC3-Discovery</td>\n",
       "      <td>drs://dg.4DFC:6301f865-15e6-11ea-9bfa-0a42f3c8...</td>\n",
       "      <td>828570494</td>\n",
       "      <td>d3e410d4c2188a45db3d0890f0329d00</td>\n",
       "      <td>Proteomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>C3N-02786</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>77cfc916-814e-48c9-8a4c-689486c7d0a4</td>\n",
       "      <td>[{'system': 'PDC', 'value': '77cfc916-814e-48c...</td>\n",
       "      <td>NCI-11Plex-23-F8-m17322.mzML.gz</td>\n",
       "      <td>Processed Mass Spectra</td>\n",
       "      <td>Open Standard</td>\n",
       "      <td>mzML</td>\n",
       "      <td>Proteogenomic Analysis of Pediatric Brain Canc...</td>\n",
       "      <td>drs://dg.4DFC:77cfc916-814e-48c9-8a4c-689486c7...</td>\n",
       "      <td>190616400</td>\n",
       "      <td>5b2f6ec7c4688f49e45ec896bbfe7eeb</td>\n",
       "      <td>Proteomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>C58917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>b1146352-9ebd-4236-8f61-baece7f6af59</td>\n",
       "      <td>[{'system': 'PDC', 'value': 'b1146352-9ebd-423...</td>\n",
       "      <td>01CPTAC_GBM_P_PNNL_20190123_B1S1_f03.mzid.gz</td>\n",
       "      <td>Peptide Spectral Matches</td>\n",
       "      <td>Open Standard</td>\n",
       "      <td>mzIdentML</td>\n",
       "      <td>CPTAC3-Discovery</td>\n",
       "      <td>drs://dg.4DFC:b1146352-9ebd-4236-8f61-baece7f6...</td>\n",
       "      <td>5394853</td>\n",
       "      <td>7f78d332dfa2daaef3452c62c934a16b</td>\n",
       "      <td>Proteomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>C3N-02256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>a9fffc9c-97cc-458d-bff4-c8fc814f3c74</td>\n",
       "      <td>[{'system': 'PDC', 'value': 'a9fffc9c-97cc-458...</td>\n",
       "      <td>NCI-11pex-8H-Phos-JM05257.mzid.gz</td>\n",
       "      <td>Peptide Spectral Matches</td>\n",
       "      <td>Open Standard</td>\n",
       "      <td>mzIdentML</td>\n",
       "      <td>Proteogenomic Analysis of Pediatric Brain Canc...</td>\n",
       "      <td>drs://dg.4DFC:a9fffc9c-97cc-458d-bff4-c8fc814f...</td>\n",
       "      <td>3171775</td>\n",
       "      <td>de574a09f9e8b76a73dd6f1fb4edc2c9</td>\n",
       "      <td>Proteomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>C20541</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>b0a661d0-312c-4ee0-bc49-66ef7eb95294</td>\n",
       "      <td>[{'system': 'PDC', 'value': 'b0a661d0-312c-4ee...</td>\n",
       "      <td>10CPTAC_GBM_P_PNNL_20190501_B3S2_f01.psm</td>\n",
       "      <td>Peptide Spectral Matches</td>\n",
       "      <td>Text</td>\n",
       "      <td>tsv</td>\n",
       "      <td>CPTAC3-Discovery</td>\n",
       "      <td>drs://dg.4DFC:b0a661d0-312c-4ee0-bc49-66ef7eb9...</td>\n",
       "      <td>2179691</td>\n",
       "      <td>af132b8401965079a0d3865acb150f99</td>\n",
       "      <td>Proteomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>C3N-03184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>621b178a-496a-4949-a609-a598bc6abc37</td>\n",
       "      <td>[{'system': 'PDC', 'value': '621b178a-496a-494...</td>\n",
       "      <td>NCI-11Plex-3-F4-f10331.mzML.gz</td>\n",
       "      <td>Processed Mass Spectra</td>\n",
       "      <td>Open Standard</td>\n",
       "      <td>mzML</td>\n",
       "      <td>Proteogenomic Analysis of Pediatric Brain Canc...</td>\n",
       "      <td>drs://dg.4DFC:621b178a-496a-4949-a609-a598bc6a...</td>\n",
       "      <td>159363389</td>\n",
       "      <td>55884838627eb0f3c79ae9d01f2faf40</td>\n",
       "      <td>Proteomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>C30258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>a1f5fd32-7c5e-4730-93f4-11cf4b1bfc04</td>\n",
       "      <td>[{'system': 'PDC', 'value': 'a1f5fd32-7c5e-473...</td>\n",
       "      <td>NCI-11pex-19C-Phos-c05390.mzML.gz</td>\n",
       "      <td>Processed Mass Spectra</td>\n",
       "      <td>Open Standard</td>\n",
       "      <td>mzML</td>\n",
       "      <td>Proteogenomic Analysis of Pediatric Brain Canc...</td>\n",
       "      <td>drs://dg.4DFC:a1f5fd32-7c5e-4730-93f4-11cf4b1b...</td>\n",
       "      <td>221231635</td>\n",
       "      <td>948cba7bf0a4e63d52b97966e0047d86</td>\n",
       "      <td>Proteomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>C387450</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      id  \\\n",
       "0   e5b8f4c3-efa1-412c-8769-6bd629ceff25   \n",
       "1   416a6c7e-15e5-11ea-9bfa-0a42f3c845fe   \n",
       "2   32a4267d-0cc1-4022-ae45-5efb4a29eef6   \n",
       "3   6301f865-15e6-11ea-9bfa-0a42f3c845fe   \n",
       "4   77cfc916-814e-48c9-8a4c-689486c7d0a4   \n",
       "..                                   ...   \n",
       "95  b1146352-9ebd-4236-8f61-baece7f6af59   \n",
       "96  a9fffc9c-97cc-458d-bff4-c8fc814f3c74   \n",
       "97  b0a661d0-312c-4ee0-bc49-66ef7eb95294   \n",
       "98  621b178a-496a-4949-a609-a598bc6abc37   \n",
       "99  a1f5fd32-7c5e-4730-93f4-11cf4b1bfc04   \n",
       "\n",
       "                                           identifier  \\\n",
       "0   [{'system': 'PDC', 'value': 'e5b8f4c3-efa1-412...   \n",
       "1   [{'system': 'PDC', 'value': '416a6c7e-15e5-11e...   \n",
       "2   [{'system': 'PDC', 'value': '32a4267d-0cc1-402...   \n",
       "3   [{'system': 'PDC', 'value': '6301f865-15e6-11e...   \n",
       "4   [{'system': 'PDC', 'value': '77cfc916-814e-48c...   \n",
       "..                                                ...   \n",
       "95  [{'system': 'PDC', 'value': 'b1146352-9ebd-423...   \n",
       "96  [{'system': 'PDC', 'value': 'a9fffc9c-97cc-458...   \n",
       "97  [{'system': 'PDC', 'value': 'b0a661d0-312c-4ee...   \n",
       "98  [{'system': 'PDC', 'value': '621b178a-496a-494...   \n",
       "99  [{'system': 'PDC', 'value': 'a1f5fd32-7c5e-473...   \n",
       "\n",
       "                                           label             data_category  \\\n",
       "0   09CPTAC_GBM_W_PNNL_20190501_B3S1_f24.mzid.gz  Peptide Spectral Matches   \n",
       "1       06CPTAC_GBM_P_PNNL_20190306_B2S2_f12.raw          Raw Mass Spectra   \n",
       "2   08CPTAC_GBM_W_PNNL_20190306_B2S4_f18.mzML.gz    Processed Mass Spectra   \n",
       "3       11CPTAC_GBM_W_PNNL_20190501_B3S3_f09.raw          Raw Mass Spectra   \n",
       "4                NCI-11Plex-23-F8-m17322.mzML.gz    Processed Mass Spectra   \n",
       "..                                           ...                       ...   \n",
       "95  01CPTAC_GBM_P_PNNL_20190123_B1S1_f03.mzid.gz  Peptide Spectral Matches   \n",
       "96             NCI-11pex-8H-Phos-JM05257.mzid.gz  Peptide Spectral Matches   \n",
       "97      10CPTAC_GBM_P_PNNL_20190501_B3S2_f01.psm  Peptide Spectral Matches   \n",
       "98                NCI-11Plex-3-F4-f10331.mzML.gz    Processed Mass Spectra   \n",
       "99             NCI-11pex-19C-Phos-c05390.mzML.gz    Processed Mass Spectra   \n",
       "\n",
       "        data_type      file_format  \\\n",
       "0   Open Standard        mzIdentML   \n",
       "1     Proprietary  vendor-specific   \n",
       "2   Open Standard             mzML   \n",
       "3     Proprietary  vendor-specific   \n",
       "4   Open Standard             mzML   \n",
       "..            ...              ...   \n",
       "95  Open Standard        mzIdentML   \n",
       "96  Open Standard        mzIdentML   \n",
       "97           Text              tsv   \n",
       "98  Open Standard             mzML   \n",
       "99  Open Standard             mzML   \n",
       "\n",
       "                                   associated_project  \\\n",
       "0                                    CPTAC3-Discovery   \n",
       "1                                    CPTAC3-Discovery   \n",
       "2                                    CPTAC3-Discovery   \n",
       "3                                    CPTAC3-Discovery   \n",
       "4   Proteogenomic Analysis of Pediatric Brain Canc...   \n",
       "..                                                ...   \n",
       "95                                   CPTAC3-Discovery   \n",
       "96  Proteogenomic Analysis of Pediatric Brain Canc...   \n",
       "97                                   CPTAC3-Discovery   \n",
       "98  Proteogenomic Analysis of Pediatric Brain Canc...   \n",
       "99  Proteogenomic Analysis of Pediatric Brain Canc...   \n",
       "\n",
       "                                              drs_uri  byte_size  \\\n",
       "0   drs://dg.4DFC:e5b8f4c3-efa1-412c-8769-6bd629ce...    8445510   \n",
       "1   drs://dg.4DFC:416a6c7e-15e5-11ea-9bfa-0a42f3c8...  615428786   \n",
       "2   drs://dg.4DFC:32a4267d-0cc1-4022-ae45-5efb4a29...  200563993   \n",
       "3   drs://dg.4DFC:6301f865-15e6-11ea-9bfa-0a42f3c8...  828570494   \n",
       "4   drs://dg.4DFC:77cfc916-814e-48c9-8a4c-689486c7...  190616400   \n",
       "..                                                ...        ...   \n",
       "95  drs://dg.4DFC:b1146352-9ebd-4236-8f61-baece7f6...    5394853   \n",
       "96  drs://dg.4DFC:a9fffc9c-97cc-458d-bff4-c8fc814f...    3171775   \n",
       "97  drs://dg.4DFC:b0a661d0-312c-4ee0-bc49-66ef7eb9...    2179691   \n",
       "98  drs://dg.4DFC:621b178a-496a-4949-a609-a598bc6a...  159363389   \n",
       "99  drs://dg.4DFC:a1f5fd32-7c5e-4730-93f4-11cf4b1b...  221231635   \n",
       "\n",
       "                            checksum data_modality imaging_modality  \\\n",
       "0   b707f6f6abb880553146af8214911d06     Proteomic             None   \n",
       "1   e2620190ade98fff9bffd0149feefd8d     Proteomic             None   \n",
       "2   170fca4126b7a4977c5d2248f0b9b537     Proteomic             None   \n",
       "3   d3e410d4c2188a45db3d0890f0329d00     Proteomic             None   \n",
       "4   5b2f6ec7c4688f49e45ec896bbfe7eeb     Proteomic             None   \n",
       "..                               ...           ...              ...   \n",
       "95  7f78d332dfa2daaef3452c62c934a16b     Proteomic             None   \n",
       "96  de574a09f9e8b76a73dd6f1fb4edc2c9     Proteomic             None   \n",
       "97  af132b8401965079a0d3865acb150f99     Proteomic             None   \n",
       "98  55884838627eb0f3c79ae9d01f2faf40     Proteomic             None   \n",
       "99  948cba7bf0a4e63d52b97966e0047d86     Proteomic             None   \n",
       "\n",
       "   dbgap_accession_number imaging_series researchsubject_specimen_id  \\\n",
       "0                    None           None                               \n",
       "1                    None           None                               \n",
       "2                    None           None                               \n",
       "3                    None           None                               \n",
       "4                    None           None                               \n",
       "..                    ...            ...                         ...   \n",
       "95                   None           None                               \n",
       "96                   None           None                               \n",
       "97                   None           None                               \n",
       "98                   None           None                               \n",
       "99                   None           None                               \n",
       "\n",
       "   researchsubject_id subject_id  \n",
       "0                      C3N-01366  \n",
       "1                      C3N-02185  \n",
       "2                      C3N-01856  \n",
       "3                      C3N-02786  \n",
       "4                         C58917  \n",
       "..                ...        ...  \n",
       "95                     C3N-02256  \n",
       "96                        C20541  \n",
       "97                     C3N-03184  \n",
       "98                        C30258  \n",
       "99                       C387450  \n",
       "\n",
       "[100 rows x 17 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fileresults = myquery.file.run()\n",
    "fileresults.to_dataframe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46fcb3e4",
   "metadata": {},
   "source": [
    "As you might expect, searching file gives us a huge number of results. This is great if you are surveying what kind of data is available, but is less useful for getting a coherent cohort. \n",
    "\n",
    "A better way to get files for a specific cohort is to chain your queries together, which we cover in the next tutorial <a href=\"../Chaining\">Chaining Endpoints</a>: Combine information from multiple endpoints, and build And/Or/Like and other advanced query strings.\n",
    "\n",
    "Another useful way to look at high level information is to use our counts feature which returns summary information rather than the full search results. Check out the <a href=\"../DataSummaries\">Data Summaries tutorial</a> to try it.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "549257b1",
   "metadata": {},
   "source": [
    "\n",
    "---\n",
    "\n",
    "<div class=\"cdadefine\" style=\"background-color:#add9e5;color:black;padding:20px;\">\n",
    "\n",
    "<h3>File Field Definitions</h3>\n",
    "\n",
    "<i>A file is an information-bearing electronic object that contains a physical embodiment of some information using a particular character encoding.</i>\n",
    "\n",
    "    \n",
    "<ul>\n",
    "  <li><b>id:</b> The 'logical' identifier of the entity in the system of record, e.g. a UUID. This 'id' is unique within a given system. The identified entity may have a different 'id' in a different system.</li>\n",
    "<li><b>identifier:</b> A 'business' identifier for the entity, typically as provided by an external system or authority, that persists across implementing systems (i.e. a 'logical' identifier). Uses a specialized, complex 'Identifier' data type to capture information about the source of the business identifier - or a URI expressed as a string to an existing entity.</li>\n",
    "<li><b>identifier.system:</b> The system or namespace that defines the identifier.</li>\n",
    "<li><b>identifier.value:</b> The value of the identifier, as defined by the system.</li>\n",
    "  <li><b>label:</b> Short name or abbreviation for dataset. Maps to rdfs:label.</li>\n",
    "  <li><b>data_catagory:</b> Broad categorization of the contents of the data file.</li>\n",
    "  <li><b>data_type:</b> Specific content type of the data file.</li>\n",
    "  <li><b>file_format:</b> Format of the data files.</li>\n",
    "  <li><b>associated_project:</b> A reference to the Project(s) of which this ResearchSubject is a member. The associated_project may be embedded using the ref definition or may be a reference to the id for the Project - or a URI expressed as a string to an existing entity.</li>\n",
    "  <li><b>drs_uri:</b> A string of characters used to identify a resource on the Data Repo Service(DRS). Can be used to retreive this specific file from a server.</li>\n",
    "  <li><b>byte_size:</b> Size of the file in bytes. Maps to dcat:byteSize.</li>\n",
    "  <li><b>checksum:</b> The md5 value for the file. A digit representing the sum of the correct digits in a piece of stored or transmitted digital data, against which later comparisons can be made to detect errors in the data.</li>\n",
    "  <li><b>data_modality:</b> Data modality describes the biological nature of the information gathered as the result of an Activity, independent of the technology or methods used to produce the information. Always one of \"Genomic\", \"Proteomic\", or \"Imaging\".</li>\n",
    "  <li><b>imaging_modality:</b> An imaging modality describes the imaging equipment and/or method used to acquire certain structural or functional information about the body. These include but are not limited to computed tomography (CT) and magnetic resonance imaging (MRI). Taken from the DICOM standard.</li>\n",
    "  <li><b>dbgap_accession_number:</b> The dbgap accession number for the project.</li>\n",
    "</ul>  \n",
    "\n",
    "</div>\n",
    "    \n",
    "---"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "5c867d1980d5b66d2bfc8a5903dcee074b4b68f3917a4f27f8a310cab24e9f1b"
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.13"
  },
  "metadata": {
   "interpreter": {
    "hash": "5c867d1980d5b66d2bfc8a5903dcee074b4b68f3917a4f27f8a310cab24e9f1b"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
